proposals/function-pointers.md

```diff
@@ -0,0 +1,245 @@
+- Custom attributes cannot be applied to a `funcptr` or any of its elements.
+- A `funcptr` parameter cannot be marked as `params`
+- A `funcptr` type has all of the restrictions of a normal pointer type.
+
+### Allow addresss-of to target methods
+
+Method groups will now be allowed as arguments to an address-of expression. The type of such an 
+expression will be an unnamed `funcptr` which has the equivalent signature of the target method and a managed 
+calling convention:
+
+``` csharp
+unsafe class Util { 
+    public static void Log() { } 
+
+    funcptr void Action();
+    funcptr int Func();
+    void Use() {
+        funcptr void() ptr1 = &Util.Log; 
+        Action ptr2 = &Util.Log;
+
+        // Error: type "funcptr void()" not compatible with "funcptr int()";
+        Func ptr3 = &Util.Log; 
+
+        // Okay. Conversion to void* is always allowed.
+        void* v = &Util.Log;
+   }
+}
+```
+
+The conversion of an address-of method group to `funcptr` has roughly the same process as method group to `delegate`  
+conversion. There are two additional restrictions to the existing process:
+- Only members of the method group that are marked as `static` will be considered. 
+- Only a `funcptr` with a managed calling convention can be the target of such a conversion.
+
+This means developers can depend on overload resolution rules to work in conjunction with the 
+address-of operator:
+
+``` csharp
+unsafe class Util { 
+    public static void Log() { } 
+    public static void Log(string p1) { } 
+    public static void Log(int i) { };
+
+    funcptr void Action1();
+    funcptr void Action2();
+
+    void Use() {
+        Action1 a1 = &Log; // Log()
+        Action2 a2 = &Log; // Log(int i)
+
+        // Error: ambiguous conversion from method group Log to "void*"
+        void* v = &Log; 
+    }
+```
+
+The address-of operator will be implemented using the `ldftn` instruction.
+
+Restrictions of this feature:
+- Only applies to methods marked as `static`.
+- Local functions cannot be used in `&`. The implementation details of these methods are
+deliberately not specified by the language. This includes whether they are static vs. instance or
+exactly what signature they are emitted with.
+
+### Better function member
+The better function member specification will be changed to include the following line:
+
+> A `funcptr` is more specific than `void*`
+
+This means that it is possible to overload on `void*` and a `funcptr` and still sensibly use the address-of operator.
+
+## Open Issuess
+
+- Round tripping function pointer names, as well as parameter names, through metadata will require additional work. The
+function pointer type itself is natively supported by CLI but that does not include any names. This is not anticipated
+to be a big issue, just needs design work.
+- The address-of operator is limited to `static` methods in this proposal. It can be made to work with instance methods 
```

Comment:
FWIW, this would not always work. Static methods and instance methods do not have the same calling convention. static function pointer to instance method cannot be manufactured by pre-pending "this" pointer.

Reply:
Where can I get some more info on how this does and doesn't work?

Reply:
@jkotas 

> FWIW, this would not always work. Static methods and instance methods do not have the same calling convention. static function pointer to instance method cannot be manufactured by pre-pending "this" pointer.

Why not? What are the cases where it would not work?

Reply:
Methods with return buffers. For example:

```csharp
struct BigStruct { long a,b,c,d,e,f; }

class MyClass {
    BigStruct MyMethod();
}
```

On Windows x64, `MyMethod` takes `this` in `rcx` and return buffer pointer in `rdx`.

Static `BigStruct pfn(MyClass)` function pointed takes return buffer in `rcx` and `MyClass` argument in `rdx`.

Reply:
@jkotas Thanks for the explanation.